### PR TITLE
Optimise template `_d_arrayappendT`

### PR DESCRIPTION
**Pull request by teodutu**
The current implementation is slow because it calls `copyEmplace` on every new element of the destination array. If `T` has a postblit, it is more efficient to blit all new elements at once and only then call `__xpostblit()` on them individually.

### CODE REVIEW

**Comment by maxhaton** [comment]
> Faster

Where's the benchmarking data?

**Comment by RazvanN7** [comment]
@maxhaton Benchmarking is important to know exactly what the benefits are, however, in this case I would say that it is obviously better to do a single mempcy instead of more.

**Comment by dkorpel** [comment]
Is this Exception safe?

**Comment by RazvanN7** [comment]
What do you mean by exception?

**Comment by dkorpel** [comment]
> What do you mean by exception?

If one of the postblits throws an Exception, will cleanup happen properly?

**Comment by teodutu** [comment]
@maxhaton Below are the running times of concatenating a 256-element array to an empty array 1M times:
| Struct size | Old Hook Time (ms) | New Hook Time (ms) | Performance Increase |
|-------------|:------------------:|:------------------:|:--------------------:|
| 0 bytes     | 1442.1             | 440.29             | 69.4%                |
| 64 bytes    | 1902.85            | 1109.18            | 41.7%                |
| 256 bytes   | 4393.38            | 2978.35            | 32.2%                |

Yes, the data was collected using dmd because the new lowering to `_d_arrayappendT` is not integrated in ldc / gdc yet.

**Comment by RazvanN7** [comment]
@teodutu what does old hook mean? The un-templatised one or the templatised one without this optimization?

**Comment by teodutu** [comment]
@dkorpel 
> If one of the postblits throws an Exception, will cleanup happen properly?

Then the hook will throw the Exception to the calling function. The same thing happens now when using `copyEmplace`, because it calls `__xpostblit` itself: https://github.com/dlang/druntime/blob/a63ae87f63b2fba058241ad410b9c1b5b43438d2/src/core/lifetime.d#L1266-L1270

**Comment by teodutu** [comment]
@RazvanN7 It refers to the old **template** hook. I've updated my comment. The performance of the new template hook is the same as that of the older, non-template hook.

**Comment by maxhaton** [comment]
> @maxhaton Below are the running times of concatenating a 256-element array to an empty array 1M times:
> 
> Struct size	Old Template Hook Time (ms)	New Template Hook Time (ms)	Performance Increase
> 0 bytes	1442.1	440.29	69.4%
> 64 bytes	1902.85	1109.18	41.7%
> 256 bytes	4393.38	2978.35	32.2%
> Yes, the data was collected using dmd because the new lowering to `_d_arrayappendT` is not integrated in ldc / gdc yet.

You don't need the lowering actually in ldc or GDC, surely.  You know what AST it's going to generate.

You might be able to copy and paste some of the AST output directly into a file if you're really lucky.

**Comment by maxhaton** [comment]
And the data is promising but it's important to measure because a few times I've tried to do things like this only to find ldc or GDC either don't use the hooks at all or already generated what i was trying to do.

**Review by RazvanN7 — APPROVED** [review]
Looks good to me. @dkorpel @maxhaton Do you request further information or updates for this PR or is this good to go?

**Comment by dkorpel** [comment]
> Then the hook will throw the Exception to the calling function.

Yes, but at that point you already have a memcpy of the rest of the array, which will have their destructor called without postblit, which could lead to double frees. However, this seems to be an issue with the current implementation as well:
```D
import core.stdc.stdio;
import core.memory;

int g = 0;

struct S {
    int* ptr;
    
    this(int x) {
        printf("create %d\n", x);
        this.ptr = new int(x);
    }
    
    this(this) {
        if (g++ > 2) {
            throw new Exception("");
        }
        ptr = new int(*ptr + 1);
    }
    
    ~this() {printf("free %d %p\n", *ptr, ptr);}
}

void f() {
    try {
        S[] a = [S(10), S(20)];
        a ~= [S(30), S(40), S(50)];
    } catch (Exception e) {
        printf("EXCEPTION\n");
    }
}

void main()
{
    f();
    GC.collect();
}
```

```
create 10
create 20
create 30
create 40
create 50
EXCEPTION
free 20 0x7fb3b4ecf010
free 10 0x7fb3b4ecf000
free 50 0x7fb3b4ecf040
free 40 0x7fb3b4ecf030
free 30 0x7fb3b4ecf020
free 50 0x7fb3b4ecf040 <-- double free on blitted but not postblitted instance
free 40 0x7fb3b4ecf030 <-- ditto
free 31 0x7fb3b4ecf070 
free 21 0x7fb3b4ecf060
free 11 0x7fb3b4ecf050
```

File: src/core/internal/traits.d [hunk 1]
```diff
@@ -274,6 +274,23 @@ template hasElaborateDestructor(S)
     }
 }
 
+// std.traits.hasElaboratePostblit
```
Comment (dkorpel) [inline src/core/internal/traits.d:277]: I don't see `hasElaboratePostblit` in std.traits, what does this mean? 

A full DDoc comment would be appreciated.

File: src/core/internal/traits.d [hunk 2]
```diff
@@ -274,6 +274,23 @@ template hasElaborateDestructor(S)
...
+template hasElaboratePostblit(S)
+{
+    static if (__traits(isStaticArray, S))
+    {
+        enum bool hasElaboratePostblit = S.sizeof && hasElaboratePostblit!(BaseElemOf!S);
+    }
+    else static if (is(S == struct))
```
Comment (dkorpel) [inline src/core/internal/traits.d:284]: What about `S == enum` cases?

File: src/core/internal/traits.d [hunk 3]
```diff
@@ -274,6 +274,23 @@ template hasElaborateDestructor(S)
     }
 }
 
+// std.traits.hasElaboratePostblit
+template hasElaboratePostblit(S)
```
Comment (dkorpel) [inline src/core/internal/traits.d:278]: What's 'elaborate' about this? Looks like `hasPostblit` to me.

**Review by dkorpel — COMMENTED** [review]
Considering the postblit is superseded by the copy constructor, I don't think it's useful adding complexity with `hasElaboratePostblit`, unless this is motivated by a benchmark of a real code base.

File: src/core/internal/traits.d (hunk 3 above)
Comment (teodutu) [reply src/core/internal/traits.d]: You're right. I mistakenly believed `hasPostblit` only looks for a postblit if the type is a struct and is false when the type is an array.

**Comment by teodutu** [comment]
@dkorpel You were right that `hasElaboratePostblit` was useless. Therefore, I deleted it and used `hasPostblit` instead.

About the change to `_d_arrayctor`, I believe it's useful because postblits are not deprecated yet. Therefore, it's counterproductive to slow down existing code when a speedup is evident.

You're also right about throwing exceptions possibly causing double frees. However, as you noted, this issue is present in the current template `_d_arrayappendT` as well as in its old non-template counterpart [1]. Therefore I believe this work should have a separate PR as it's a fix, rather than an optimisation.

[1] https://github.com/dlang/druntime/blob/bdeee862aaa125e75aebc3118a43d2d5e5ee954d/src/rt/lifetime.d#L1878-L1890

**Review by dkorpel — APPROVED** [review]

**Comment by RazvanN7** [comment]
@teodutu could you maybe rebase this? That will restart the documentation tester which seems to have a network related issue.

**Pull request merged**